### PR TITLE
feat(@aws-amplify/ui-components): amplify-s3-image component

### PR DESCRIPTION
**Pull request by jordanranz**
_Description of changes:_
- `amplify-s3-image` component
 
Validation steps:
- From `amplify-js`, `cd packages/amplify-ui-components`
- Copy or update `aws-exports.js` file with new storage config, ping me if need be
- Run `yarn` then `yarn start`
- Storage example should show up first, if not navigate to Storage sample

Usage:
`<amplify-s3-image imgKey="amplify-logo.png">`
- `imgKey` represents the key of the image inside the Amplify CLI configured S3 bucket

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-ui-components/src/components/amplify-s3-image/amplify-s3-image.tsx [hunk 1]
```diff
@@ -21,7 +21,7 @@ export class AmplifyS3Image {
   @Prop() handleOnLoad: () => {};
   @Prop() handleOnError: () => {};
 
-  @State() src: string = 'https://www.systutorials.com/wp/files/2015/11/AmazonS3-e1448589113191.png';
+  @State() src: string = null;
```
Comment (sammartinez) [inline packages/amplify-ui-components/src/components/amplify-s3-image/amplify-s3-image.tsx]: Does it make sense to have this be set to `null`, would `undefined` be the same here?

File: packages/amplify-ui-components/src/components/amplify-examples/amplify/backend/auth/amplifyexamples15aa6dfd/amplifyexamples15aa6dfd-cloudformation-template.yml [hunk 2]
```diff
@@ -0,0 +1,364 @@
+AWSTemplateFormatVersion: 2010-09-09
```
Comment (sammartinez) [inline packages/amplify-ui-components/src/components/amplify-examples/amplify/backend/auth/amplifyexamples15aa6dfd/amplifyexamples15aa6dfd-cloudformation-template.yml]: Is this file needed for the code or can we remove them ?

File: packages/amplify-ui-components/src/components/amplify-examples/amplify/backend/auth/amplifyexamples15aa6dfd/parameters.json [hunk 3]
```diff
@@ -0,0 +1,50 @@
+{
```
Comment (sammartinez) [inline packages/amplify-ui-components/src/components/amplify-examples/amplify/backend/auth/amplifyexamples15aa6dfd/parameters.json]: Same with this file

File: packages/amplify-ui-components/src/components/amplify-examples/amplify/backend/backend-config.json [hunk 4]
```diff
@@ -1,15 +1,26 @@
...
-    }
-  },
-  "xr": {
-    "classroom": {
-      "service": "Sumerian",
-      "providerPlugin": "awscloudformation"
-    }
-  }
-}
+	"auth": {
```
Comment (sammartinez) [inline packages/amplify-ui-components/src/components/amplify-examples/amplify/backend/backend-config.json]: Same with this file

File: packages/amplify-ui-components/src/components/amplify-examples/amplify/backend/storage/s37d218eea/parameters.json [hunk 5]
```diff
@@ -0,0 +1,37 @@
+{
```
Comment (sammartinez) [inline packages/amplify-ui-components/src/components/amplify-examples/amplify/backend/storage/s37d218eea/parameters.json]: same with this file

File: packages/amplify-ui-components/src/components/amplify-s3-image/amplify-s3-image.tsx [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+  @Prop() path: string;
+  /* Image body content to be uploaded */
+  @Prop() body: object;
+  /* The content type header used when uploading to S3 */
+  @Prop() contentType: string = 'binary/octet-stream';
+  /* The accessbility level of the image */
+  @Prop() level: string;
+  /* Whether or not to use track the get/put of the image */
+  @Prop() track: boolean;
+  /* Cognito ideneity id of the another user's image */
```
Comment (sammartinez) [inline packages/amplify-ui-components/src/components/amplify-s3-image/amplify-s3-image.tsx]:
```suggestion
  /* Cognito identity id of the another user's image */
```

File: packages/amplify-ui-components/src/components/amplify-s3-image/amplify-s3-image.tsx [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+  @Prop() identityId: string;
+  /* Whether or not the photo picker is enabled */
+  @Prop() pickerEnabled: boolean = true;
+  /** Override default styling */
+  @Prop() overrideStyle: boolean = false;
+  /* Function executed when image loads */
+  @Prop() handleOnLoad: () => {};
```
Comment (sammartinez) [inline packages/amplify-ui-components/src/components/amplify-s3-image/amplify-s3-image.tsx]: Do we want these to void?

File: packages/amplify-ui-components/src/components/amplify-s3-image/amplify-s3-image.tsx [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  async getImageSource(key, level, track, identityId) {
+    if (!Storage || typeof Storage.get !== 'function') {
+      throw new Error(NO_STORAGE_MODULE_FOUND);
+    }
+
+    try {
+      const src = await Storage.get(key, { level: level ? level : 'public', track, identityId });
+      return src;
```
Comment (sammartinez) [inline packages/amplify-ui-components/src/components/amplify-s3-image/amplify-s3-image.tsx]: Would the below suggestion work? 
```suggestion
      await Storage.get(key, { level: level ? level : 'public', track, identityId });
      return;
```

I guess either is fine cause ES linter recommends both so

**Comment by ericclemmons** [comment]
How would you recommend we validate the functionality of this PR?

File: packages/amplify-ui-components/src/components/amplify-s3-image/amplify-s3-image.tsx (hunk 8 above)
Comment (jordanranz) [reply packages/amplify-ui-components/src/components/amplify-s3-image/amplify-s3-image.tsx]: The suggestion would not work because the anything consuming `getImageSource` expects it to return a string

**Comment by jordanranz** [comment]
@ericclemmons The best way would probably be to clone my fork, hop on this branch and start the stencil example server.

File: packages/amplify-ui-components/src/components/amplify-examples/amplify/backend/auth/amplifyexamples15aa6dfd/amplifyexamples15aa6dfd-cloudformation-template.yml (hunk 2 above)
Comment (jordanranz) [reply packages/amplify-ui-components/src/components/amplify-examples/amplify/backend/auth/amplifyexamples15aa6dfd/amplifyexamples15aa6dfd-cloudformation-template.yml]: This file is needed

File: packages/amplify-ui-components/src/components/amplify-examples/amplify/backend/auth/amplifyexamples15aa6dfd/parameters.json (hunk 3 above)
Comment (jordanranz) [reply packages/amplify-ui-components/src/components/amplify-examples/amplify/backend/auth/amplifyexamples15aa6dfd/parameters.json]: This file is needed

File: packages/amplify-ui-components/src/components/amplify-examples/amplify/backend/backend-config.json (hunk 4 above)
Comment (jordanranz) [reply packages/amplify-ui-components/src/components/amplify-examples/amplify/backend/backend-config.json]: This file is needed

**Comment by ericclemmons** [comment]
Sorry, I should've been clearer.
> How would you recommend we validate the functionality of this PR?

What I meant was, _how would you recommend we validate **correctness** of this PR?_

There are several props, cloudformation templates, auth stuff, and more, so understanding what it should and shouldn't do is crucial.

Here are a couple suggestions that I've found extremely helpful for teams in the past:

1. Step-by-step "getting started" instructions.

	In my case, I ran `cd packages/amplify-ui-components`, `yarn`, then `yarn start`, but am greeted by this error and a blank screen:

	```
  	[ERROR] 28:10.642 S3Image Error: "No input provided for required configuration parameter: region"
  	```

	`ui-components/master` works as expected.
  

1. End-user validation steps.

	For example, "selecting an image should show a preview", "file picker should only show PNG/BMP/JPG extensions", "uploading an image > 10mb should show an error", etc.

With that in mind, is there anything particular you'd like for validation, or only code review?

**Comment by jordanranz** [comment]
Ok, I understand your question better now.

> how would you recommend we validate correctness of this PR?

I would say that the correctness is determined by the functionality exhibited by the [documentation of the most mature framework (React)](https://aws-amplify.github.io/docs/js/storage#ui-components-for-react). Most of the props passed into this component are loosely documented there. For the remainder I'd check the auto generated comments for the Stencil component.

As for the cloudformation changes. These are simply artifacts created when initiating `amplify add storage` to our test account so that we can interact with the Storage category from our example app.

I like both of the approaches mentioned and I think it is valuable to include these type of validations to the PR. The question is how verbose should these validation steps be?

I will be adding testing/storybook today for more clarity here.

File: packages/amplify-ui-components/src/components/amplify-s3-image/amplify-s3-image.tsx (hunk 7 above)
Comment (jordanranz) [reply packages/amplify-ui-components/src/components/amplify-s3-image/amplify-s3-image.tsx]: Changed this to match the correct event function type.

File: packages/amplify-ui-components/src/components/amplify-s3-image/amplify-s3-image.tsx [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+  /* Callback used to generate custom key value */
+  @Prop() fileToKey: (data: object) => string;
+  /* Function executed when image loads */
+  @Prop() handleOnLoad: (event: Event) => void;
+  /* Function executed when error occurs for the image */
+  @Prop() handleOnError: (event: Event) => void;
+
+  @State() src: string | object;
+
+  async componentWillLoad() {
```
Comment (sammartinez) [inline packages/amplify-ui-components/src/components/amplify-s3-image/amplify-s3-image.tsx]: NIT: I am seeing that the React implementation has a `componentDidMount` instead of the `componentWillMount` which is now marked as unsafe. Was there a reason to use the Stencil `componentWillLoad` over the [componentDidLoad](https://stenciljs.com/docs/component-lifecycle#componentdidload-)

File: packages/amplify-ui-components/src/components/amplify-s3-image/amplify-s3-image.tsx [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+
+  async load() {
+    const { imgKey, path, body, contentType, level, track, identityId } = this;
+    if (!imgKey && !path) {
+      logger.debug('empty imgKey and path');
+      return;
+    }
+
+    const key = imgKey || path;
+    logger.debug('loading ' + key + '...');
```
Comment (sammartinez) [inline packages/amplify-ui-components/src/components/amplify-s3-image/amplify-s3-image.tsx]: Can we make this a string literal instead? 

```suggestion
    logger.debug(`loading ${key}...`);
```

File: packages/amplify-ui-components/src/components/amplify-s3-image/amplify-s3-image.tsx (hunk 9 above)
Comment (jordanranz) [reply packages/amplify-ui-components/src/components/amplify-s3-image/amplify-s3-image.tsx]: Spoke offline about this.

File: packages/amplify-ui-components/src/components/amplify-s3-image/amplify-s3-image.tsx (hunk 10 above)
Comment (jordanranz) [reply packages/amplify-ui-components/src/components/amplify-s3-image/amplify-s3-image.tsx]: yep. good call.
